### PR TITLE
feat: enable talos client to connect to Talos through an auth proxy

### LINKED ISSUE #5980: Talos API access through an authentication proxy
- server CA validation: might be nothing, if the server CA is publicly trusted
- several client authentication schemes:
   - (as it is today): client cert + key
   - Basic Auth
   - no auth (for development purposes)
   - other types auth (TBD)
 - optional `cluster` field to target a specific cluster behind a proxy
 - endpoint becomes endpoint of the authentication proxy

`talosconfig` today:

```yaml
    talos-default-457:
        endpoints:
            - 172.20.0.2
        ca: LS0tLS1CRUdJTiBDRV
        crt: LS0tLS1CRUd
        key: LS0t
```

`talosconfig` for going through an auth proxy:

```yaml
    talos-via-proxy:
        endpoints:
            - [https://]auth.proxy:453  # scheme is optional, `grpc://` for HTTP/2 h2c
        ca: LS0tLS1CRUdJTiBDRV # optional
        auth:
           basic:
              username: XYZ
              password: UIO
        cluster: my-lovely-cluster-1
```

`talosctl` new flags:

- `--cluster` flag

### PR DESCRIPTION
**Pull request by sauterp**
# Pull Request
Talos client can connect to Talos API via a proxy with basic auth.
Additionally it is now optional to specify a TLS CA,key or crt. Optionally
Developers can build talosctl with WITH_DEBUG=1 to allow insecure
connections when http:// endpoints are specified.

Fixes #5980

### CODE REVIEW

File: pkg/machinery/client/connection.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	tlsConfig = &tls.Config{}
+
+	if err := c.resolveConfigContext(); err != nil {
+		return nil, fmt.Errorf("failed to resolve configuration context: %w", err)
+	}
+
+	basicAuth := c.options.configContext.Auth.Basic
+	if basicAuth != nil {
+		dialOpts = append(dialOpts, WithGRPCBasicAuth(basicAuth.Username, basicAuth.Password))
+		if shouldInsecureConnectionsBeAllowed(endpoints) {
```
Comment (smira) [inline pkg/machinery/client/connection.go:63]: why are we taking a shortcut here? 

we set the basic auth, and that should be it.

whether to use insecure or TLS-based connection should be decided below

File: pkg/machinery/client/connection.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+			return c.makeConnection(ctx, target, insecure.NewCredentials(), dialOpts)
+		}
+	}
+
+	caBytes, err := getCA(c.options.configContext)
```
Comment (smira) [inline pkg/machinery/client/connection.go:68]: lines 68-91 here seems to be building `credentials.TransportCredentials` and we can extract this to a separate (testable) function. 

in this function we can also return `Insecure` creds if it's a debug build and insecure creds were requested

File: pkg/machinery/client/grpc_connection_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+package client
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type GRPCConnectionWrapper struct {
```
Comment (smira) [inline pkg/machinery/client/grpc_connection_wrapper.go:14]: we can embed the connection into a struct

```go
type grpcConnectionWrapper struct {
   *grpc.ClientConn
   
    clusterName string
}
```

this way all methods of `ClientConn` will be exposed via the wrapper by default, but we can override the methods we need (`Invoke` & `NewStream`). This way we can drop `Close`

also this struct shouldn't be public, as we don't need it in the  external API

File: pkg/machinery/client/grpc_connection_wrapper.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	clusterName string
+
+	conn *grpc.ClientConn
+}
+
+func NewGRPCConnectionWrapper(clusterName string, conn *grpc.ClientConn) *GRPCConnectionWrapper {
```
Comment (smira) [inline pkg/machinery/client/grpc_connection_wrapper.go:20]:
```suggestion
func newGRPCConnectionWrapper(clusterName string, conn *grpc.ClientConn) *GRPCConnectionWrapper {
```

File: pkg/machinery/client/grpc_connection_wrapper.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	ctx = metadata.AppendToOutgoingContext(ctx, "runtime", "Talos")
+
+	if c.clusterName != "" {
+		ctx = metadata.AppendToOutgoingContext(ctx, "context", c.clusterName)
+	}
+
+	return ctx
+}
+
+func (c *GRPCConnectionWrapper) Close() error {
```
Comment (smira) [inline pkg/machinery/client/grpc_connection_wrapper.go:48]: drop it (see above)

**Comment by smira** [comment]
as an overall thought: we need more tests

File: cmd/talosctl/cmd/talos/root.go [hunk 6]
```diff
@@ -94,6 +97,15 @@ func WithClient(action func(context.Context, *client.Client) error) error {
 				return fmt.Errorf("nodes are not set for the command: please use `--nodes` flag or configuration file to set the nodes to run the command against")
 			}
 
+			if Cluster == "" {
```
Comment (smira) [inline cmd/talosctl/cmd/talos/root.go:100]: this should go to `WithClientNoNodes`

File: cmd/talosctl/cmd/talos/root.go [hunk 7]
```diff
@@ -94,6 +97,15 @@ func WithClient(action func(context.Context, *client.Client) error) error {
...
+				configContext := c.GetConfigContext()
+				if configContext == nil {
+					return configContextErr
+				}
+
+				Cluster = configContext.Cluster
```
Comment (smira) [inline cmd/talosctl/cmd/talos/root.go:106]: this `Cluster` varialbe if set should go to `client.New` as an option: `WithCluster(Cluster)`.

File: pkg/machinery/client/connection.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+func (c *Client) makeConnection(ctx context.Context, target string, creds credentials.TransportCredentials, dialOpts []grpc.DialOption) (*GRPCConnectionWrapper, error) {
+	dialOpts = append(dialOpts,
+		grpc.WithTransportCredentials(creds),
+		grpc.WithInitialWindowSize(65535*32),
+		grpc.WithInitialConnWindowSize(65535*16))
+
+	conn, err := grpc.DialContext(ctx, target, dialOpts...)
+
+	return NewGRPCConnectionWrapper(c.GetConfigContext().Cluster, conn), err
```
Comment (smira) [inline pkg/machinery/client/connection.go:104]: `Cluster` should also come (as an optional override) from `Client` `Options`

File: pkg/machinery/client/connection.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+	caBytes, err := getCA(configContext)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get CA: %w", err)
+	}
+
+	if len(caBytes) > 0 {
+		tlsConfig.RootCAs = x509.NewCertPool()
+		tlsConfig.ClientCAs = x509.NewCertPool()
+
+		if ok := tlsConfig.ClientCAs.AppendCertsFromPEM(caBytes); !ok {
```
Comment (smira) [inline pkg/machinery/client/connection.go:83]: `.ClientAuth` and `.ClientCAs` are fields which are used by the server-side `tls.Config`, they shouldn't be set here: https://pkg.go.dev/crypto/tls#Config

File: pkg/machinery/client/client.go [hunk 10]
```diff
@@ -138,6 +124,24 @@ func (c *Client) GetEndpoints() []string {
 	return nil
 }
 
+// GetClusterName returns the client's endpoints from the override set with WithEndpoints
```
Comment (smira) [inline pkg/machinery/client/client.go:127]: comment has to be updated

**Review by smira — APPROVED** [review]
:cool:

**Review by frezbo — APPROVED** [review]

**Pull request merged**